mcf-browser/src/back-end/parse-mcf.js

```diff
@@ -0,0 +1,306 @@
+   * label is parsed.
+   * @type {Node}
+   */
+  curNode;
+  /**
+   * Current line number of the line being parsed, used for identifying location
+   * of syntax error in the mcf file.
+   * @type {number}
+   */
+
+  lineNum;
+
+  /**
+   * List of error messages regarding mcf syntax that are to be displayed to
+   * the user. A single entry of this array should be in the format of:
+   * [line number, line, error message]
+   * @type {Array<Array<String>>}
+   */
+  errors;
+  /**
+   * Create a ParseMcf object which keeps tracks of the current source node of
+   * each triple in the mcf and the provenance, which is the mcf file name.
+   * @param {string} fileName Name of the file to be parsed.
+   */
+  constructor(fileName) {
+    this.prov = fileName;
+    this.curNode = null;
+    this.lineNum = -1;
+    this.errors = [];
+  }
+
+  /**
+   * Parses a string represetnaing a comma separated list of property values
+   * from a line of an mcf file. Returns a list having either a string or
+   * <namespace, reference> pair.
+   *
+   * @param {string} propValues A comma separated list of property values.
+   * @return {Array<(string|Object)>} Array of
+   *     parsed values.
+   */
+  parsePropValues(propValues) {
+    const values = [];
+    // split propValues on commas which are not enclosed by double quotes
+    for (const propValue of propValues.split(/,(?=(?:(?:[^"]*"){2})*[^"]*$)/)) {
+      const namespace = propValue.split(':')[0].trim();
+      if (namespace in NAMESPACES) {
+        values.push({
+          'ns': namespace,
+          'ref': propValue.substring(propValue.indexOf(':') + 1).trim(),
+        });
+      } else if (propValue.split(':').length > 1 &&
+                 !namespace.startsWith('"')) {
+        this.errors.push([this.lineNum, this.line, 'unrecognized namespace']);
+        return [];
+      } else {
+        // push property value with surrounding double quotes trimmed
+        values.push(propValue.replace(/^[" ]*(.*?)[" ]*$/g, '$1'));
+      }
+    }
+    return values;
+  }
+
+  /**
+   * Sets curNode variable of the calling ParseMcf object based on the passed in
+   * parsed values of a line of mcf, given the property label for the line was
+   * 'Node'. The parsed value either is a local reference with or without the
+   * 'l' namespace, or it has a remote namespace. If the namespace is remote,
+   * then the dcid for curNode is set. Updates localNodeHash mapping to store
+   * the subject nodes to be displayed in home screen of browser.
+   *
+   * @param {Array<string|Object>} parsedValues The array of parsed values from
+   *     a line of mcf with property label of 'Node'.
+   */
+  setCurNode(parsedValues) {
+    if (parsedValues.length !== 1) {
+      this.errors.push(
+          [this.lineNum, this.line, ERROR_MESSAGES['curNode-length']]);
+      return;
+    }
+
+    // handle case: Node: localRef, which means parsedValues[0]==='localRef'
+    let nodeRef = parsedValues[0];
```

Comment:
nit: feels a bit strange to change the type of nodeRef in some cases. how about doing nodeRef = parsedValues[0] in the else:

```
  if (parsedValues[0] instanceof Object) {
    ...
  } else {
    nodeRef = parsedValues[0];
  }
```